### PR TITLE
Update the ISPyB messages sent for MotionCorrection and CtfFind Relion stages and add messages for other stages

### PR DESCRIPTION
**Pull request by d-j-hatton**
The current ISPyB messages for the MotionCorrection and CtfFind Relion stages do not contain records for the columns relating to the options with which Relion was run. These can be added in using the `RelionItOptions` object. In the future we will also need ISPyB messages for further stages of the Relion pipeline. The necessary `JobType` classes already exist for 2D and 3D classification. Small `JobType` classes are added here for auto-picking stages: either Relion's own AutoPick or crYOLO along with the relevant ISPyB commands (and the commands for the classification stages).

### CODE REVIEW

File: tests/test_project.py [hunk 1]
```diff
@@ -54,3 +95,94 @@ def test_results_collection_does_not_crash_for_an_empty_project():
...
+    assert fresh_results[0][0]["job002"][0].micrograph_number == 1
+    star_doc.write_file(corrected_star_path)
+    (
+        dials_data("relion_tutorial_data", pathlib=True)
+        / "MotionCorr"
+        / "job002"
+        / "RELION_JOB_EXIT_SUCCESS"
+    ).touch()
+    proj.load()
+    fresh_results = [p for p in proj.results.fresh]
```
Comment (Anthchirp) [inline tests/test_project.py:132]:
```suggestion
    fresh_results = list(proj.results.fresh)
```

File: src/relion/__init__.py [hunk 2]
```diff
@@ -178,34 +197,125 @@ def __init__(self):
...
                     self._cache[r[1]] = []
-                    for single_res in r[0][r[1]]:
+                for single_res in r[0][r[1]]:
+                    self._update_temp_validation_cache(
+                        r[0], r[1], single_res, curr_val_cache, results_for_validation
+                    )
+                    if end_time_not_seen_before:
                         self._cache[r[1]].append(r[0].for_cache(single_res))
                 if (r[1], r[2]) not in self._seen_before:
                     self._seen_before.append((r[1], r[2]))
```
Comment (Anthchirp) [inline src/relion/__init__.py:231]: this seems very abstract and is hard to parse. Maybe `r`s should be `namedtuple`s?

**Review by Anthchirp — APPROVED** [review]

File: src/relion/__init__.py (hunk 2 above)
Comment (d-j-hatton) [reply src/relion/__init__.py:231]: I agree that this stuff is a mess and needs a re-work. I'll have a think about whether there is a way that I can make it neater quickly but if it isn't easy I'll add an issue or something to remember to improve it in the future

Comment (d-j-hatton) [reply src/relion/__init__.py]: I replaced the tuple `r`s with `namedtuple`s as you suggested and I think it's a big improvement, thanks. I used two different `namedtuple`s as in some cases I want to store the Relion stage object along with the job directory and last completion time for that job, while in other cases I drop the stage object. I think this stuff will probably get some sort of overhaul at some point.
